File: src/api/v1/influencers.py
"""AI Influencer endpoints"""
from datetime import UTC, datetime

from fastapi import APIRouter, Depends, HTTPException, Query, Response

from src.auth.jwt_auth import CurrentUser, get_current_user
from src.core.dependencies import CharacterGeneratorServiceDep, InfluencerServiceDep
from src.core.moderation import MODERATION_PROMPT, STYLE_PROMPT
from src.models.entities import AIInfluencer, InfluencerStatus
from src.models.requests import (
    CreateInfluencerRequest,
    GeneratePromptRequest,
    UpdateSystemPromptRequest,
    ValidateMetadataRequest,
)
from src.models.responses import (
    GeneratedMetadataResponse,
    InfluencerCreateResponse,
    InfluencerResponse,
    ListInfluencersResponse,
    SystemPromptResponse,
)

router = APIRouter(prefix="/api/v1/influencers", tags=["Influencers"])


def _get_user_system_prompt(full_instructions: str | None) -> str | None:
    """Strip STYLE_PROMPT and MODERATION_PROMPT from instructions"""
    if not full_instructions:
        return full_instructions
    
    # The order of appending was: user_prompt + \n + STYLE + \n + MODERATION
    suffix = f"\n{STYLE_PROMPT}\n{MODERATION_PROMPT}"
    if full_instructions.endswith(suffix):
        return full_instructions[: -len(suffix)]
    
    return full_instructions


@router.get(
    "",
    response_model=ListInfluencersResponse,
    operation_id="listInfluencers",
    summary="List AI influencers",
    description="Retrieve paginated list of all AI influencers. Influencers are ordered by status (active, coming_soon, discontinued) with active influencers listed first. No authentication required.",
    responses={
        200: {"description": "List of influencers retrieved successfully"},
        422: {"description": "Validation error - Invalid query parameters"},
        429: {"description": "Rate limit exceeded"},
        500: {"description": "Internal server error"},
    },
)
async def list_influencers(
    limit: int = Query(default=50, ge=1, le=100, description="Number of influencers to return"),
    offset: int = Query(default=0, ge=0, description="Number of influencers to skip"),
    influencer_service: InfluencerServiceDep = None,
    response: Response = None,
):
    """\
    List all AI influencers
    
    Influencers are ordered by status: active, coming_soon, then discontinued.
    Active influencers are listed first.
    No authentication required for discovery.
    """
    influencers, total = await influencer_service.list_influencers(
        limit=limit,
        offset=offset,
    )

    # Add cache headers for browser/CDN caching (5 minutes)
    response.headers["Cache-Control"] = "public, max-age=300"

    influencer_responses = [
        InfluencerResponse(
            id=inf.id,
            name=inf.name,
            display_name=inf.display_name,
            avatar_url=inf.avatar_url,
            description=inf.description,
            category=inf.category,
            is_active=inf.is_active,
            parent_principal_id=inf.parent_principal_id,
            source=inf.source,
            system_prompt=_get_user_system_prompt(inf.system_instructions),
            created_at=inf.created_at,
        )
        for inf in influencers
    ]

    return ListInfluencersResponse(
        influencers=influencer_responses,
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get(
    "/{influencer_id}",
    response_model=InfluencerResponse,
    operation_id="getInfluencer",
    summary="Get influencer details",
    description="Retrieve detailed information about a specific AI influencer. No authentication required.",
    responses={
        200: {"description": "Influencer details retrieved successfully"},
        404: {"description": "Influencer not found"},
        422: {"description": "Validation error - Invalid influencer ID format"},
        429: {"description": "Rate limit exceeded"},
        500: {"description": "Internal server error"},
    },
)
async def get_influencer(
    influencer_id: str,
    influencer_service: InfluencerServiceDep = None,
    response: Response = None,
):
    """\
    Get specific AI influencer details
    
    No authentication required
    """
    influencer = await influencer_service.get_influencer(influencer_id)

    # Add cache headers (5 minutes)
    response.headers["Cache-Control"] = "public, max-age=300"

    return InfluencerResponse(
        id=influencer.id,
        name=influencer.name,
        display_name=influencer.display_name,
        avatar_url=influencer.avatar_url,
        description=influencer.description,
        category=influencer.category,
        is_active=influencer.is_active,
        parent_principal_id=influencer.parent_principal_id,
        source=influencer.source,
        system_prompt=_get_user_system_prompt(influencer.system_instructions),
        created_at=influencer.created_at,
    )


@router.post(
    "/generate-prompt",
    response_model=SystemPromptResponse,
    operation_id="generatePrompt",
    summary="Generate system instructions",
    description="Generate system instructions from a character concept prompt",
)
async def generate_prompt(
    request: GeneratePromptRequest,
    character_generator: CharacterGeneratorServiceDep,
):
    """Generate system instructions"""
    return await character_generator.generate_system_instructions(request.prompt)


@router.post(
    "/validate-and-generate-metadata",
    response_model=GeneratedMetadataResponse,
    operation_id="validateAndGenerateMetadata",
    summary="Validate instructions and generate metadata",
    description="Validate system instructions and generate metadata + avatar",
)
async def validate_and_generate_metadata(
    request: ValidateMetadataRequest,
    character_generator: CharacterGeneratorServiceDep,
):
    """Validate system instructions and generate metadata using AI"""
    return await character_generator.validate_and_generate_metadata(request.system_instructions)


@router.post(
    "/create",
    response_model=InfluencerCreateResponse,
    operation_id="createInfluencer",
    summary="Create a new influencer",
    description="Create a new AI influencer character",
)
async def create_influencer(
    request: CreateInfluencerRequest,
    influencer_service: InfluencerServiceDep,
    character_generator_service: CharacterGeneratorServiceDep,
):
    """Create a new AI influencer"""
    influencer = AIInfluencer(
        id=request.bot_principal_id,
        name=request.name,
        display_name=request.display_name,
        avatar_url=request.avatar_url,
        description=request.description,
        category=request.category,
        system_instructions=f"{request.system_instructions}\n{STYLE_PROMPT}\n{MODERATION_PROMPT}",
        personality_traits=request.personality_traits,
        initial_greeting=request.initial_greeting,
        suggested_messages=request.suggested_messages,
        is_active=InfluencerStatus.ACTIVE,
        is_nsfw=False,  # Enforce non-NSFW for all new characters
        parent_principal_id=request.parent_principal_id,
        source="user-created-influencer" if request.parent_principal_id else "admin-created-influencer",
        created_at=datetime.now(UTC),
        updated_at=datetime.now(UTC),
        metadata={},
    )

    created = await influencer_service.create_influencer(influencer)

    # Generate starter video prompt (not stored in DB)
    starter_video_prompt = await character_generator_service.generate_starter_video_prompt(
        display_name=created.display_name,
        system_instructions=request.system_instructions,
    )

    return InfluencerCreateResponse(
        id=created.id,
        name=created.name,
        display_name=created.display_name,
        avatar_url=created.avatar_url,
        description=created.description,
        category=created.category,
        is_active=created.is_active,
        parent_principal_id=created.parent_principal_id,
        source=created.source,
        starter_video_prompt=starter_video_prompt,
        created_at=created.created_at,
    )


@router.patch(
    "/{influencer_id}/system-prompt",
    response_model=InfluencerResponse,
    operation_id="updateSystemPrompt",
    summary="Update influencer system prompt",
    description="Update the system prompt for an AI influencer. Only the bot owner (user whose parent_principal_id matches) can perform this action.",
    responses={
        200: {"description": "System prompt updated successfully"},
        401: {"description": "Unauthorized - Missing or invalid authentication"},
        403: {"description": "Forbidden - User is not the bot owner"},
        404: {"description": "Influencer not found"},
        422: {"description": "Validation error - Invalid request data"},
        500: {"description": "Internal server error"},
    },
)
async def update_system_prompt(
    influencer_id: str,
    request: UpdateSystemPromptRequest,
    current_user: CurrentUser = Depends(get_current_user),  # noqa: B008
    influencer_service: InfluencerServiceDep = None,
):
    """Update an influencer's system prompt (bot owner only)"""
    # Get the influencer to check ownership
    influencer = await influencer_service.get_influencer(influencer_id)
    
    # Verify the current user is the bot owner
    if influencer.parent_principal_id != current_user.user_id:
        raise HTTPException(
            status_code=403,
            detail="Only the bot owner can update the system prompt",
        )
    
    # Append style and moderation prompts (same as create endpoint)
    full_system_instructions = f"{request.system_instructions}\n{STYLE_PROMPT}\n{MODERATION_PROMPT}"
    
    # Update the system prompt
    updated = await influencer_service.update_system_prompt(influencer_id, full_system_instructions)
    
    return InfluencerResponse(
        id=updated.id,
        name=updated.name,
        display_name=updated.display_name,
        avatar_url=updated.avatar_url,
        description=updated.description,
        category=updated.category,
        is_active=updated.is_active,
        parent_principal_id=updated.parent_principal_id,
        source=updated.source,
        system_prompt=_get_user_system_prompt(updated.system_instructions),
        created_at=updated.created_at,
    )


@router.delete(
    "/{influencer_id}",
    response_model=InfluencerResponse,
    operation_id="deleteInfluencer",
    summary="Soft delete influencer",
    description="Soft delete an AI influencer by marking it as discontinued and renaming to 'Deleted Bot'. Only the bot owner (user whose parent_principal_id matches) can perform this action. Chat history is preserved.",
    responses={
        200: {"description": "Influencer soft deleted successfully"},
        401: {"description": "Unauthorized - Missing or invalid authentication"},
        403: {"description": "Forbidden - User is not the bot owner"},
        404: {"description": "Influencer not found"},
        500: {"description": "Internal server error"},
    },
)
async def delete_influencer(
    influencer_id: str,
    current_user: CurrentUser = Depends(get_current_user),  # noqa: B008
    influencer_service: InfluencerServiceDep = None,
):
    """Soft delete an influencer (bot owner only)"""
    # Get the influencer to check ownership
    influencer = await influencer_service.get_influencer(influencer_id)
    
    # Verify the current user is the bot owner
    if influencer.parent_principal_id != current_user.user_id:
        raise HTTPException(
            status_code=403,
            detail="Only the bot owner can delete the influencer",
        )
    
    # Soft delete the influencer
    deleted = await influencer_service.soft_delete_influencer(influencer_id)
    
    return InfluencerResponse(
        id=deleted.id,
        name=deleted.name,
        display_name=deleted.display_name,
        avatar_url=deleted.avatar_url,
        description=deleted.description,
        category=deleted.category,
        is_active=deleted.is_active,
        parent_principal_id=deleted.parent_principal_id,
        source=deleted.source,
        system_prompt=_get_user_system_prompt(deleted.system_instructions),
        created_at=deleted.created_at,
    )

File: src/routes/influencers.rs
use std::sync::Arc;

use axum::Json;
use axum::extract::{Path, Query, State};
use axum::http::{HeaderMap, header};
use validator::Validate;

use crate::AppState;
use crate::error::{AppError, ErrorBody};
use crate::middleware::AuthenticatedUser;
use crate::models::entities::{AIInfluencer, InfluencerStatus};
use crate::models::requests::{
    CreateInfluencerRequest, GeneratePromptRequest, GenerateVideoPromptRequest, PaginationParams,
    UpdateSystemPromptRequest, ValidateMetadataRequest,
};
use crate::models::responses::{
    GeneratedMetadataResponse, InfluencerResponse, ListInfluencersResponse,
    ListTrendingInfluencersResponse, SystemPromptResponse, TrendingInfluencerResponse,
    VideoPromptResponse,
};
use crate::services::character_generator::CharacterGeneratorService;
use crate::services::moderation;

/// Fetch profile picture from User Info Service canister for main user accounts
async fn fetch_user_profile_pic(agent: &ic_agent::Agent, principal_id: &str) -> Option<String> {
    let principal = candid::Principal::from_text(principal_id).ok()?;

    let canister_id = yral_canisters_client::ic::USER_INFO_SERVICE_ID;
    let service = yral_canisters_client::user_info_service::UserInfoService(canister_id, agent);

    match service.get_users_profile_details(vec![principal]).await {
        Ok(yral_canisters_client::user_info_service::Result9::Ok(details)) => details
            .into_iter()
            .next()
            .and_then(|d| d.profile_picture.map(|p| p.url)),
        Ok(yral_canisters_client::user_info_service::Result9::Err(e)) => {
            tracing::warn!(error = %e, "Canister error fetching user profile");
            None
        }
        Err(e) => {
            tracing::warn!(error = %e, "IC agent error fetching user profile");
            None
        }
    }
}

/// Fetch username from metadata server for main user accounts
async fn fetch_username_from_metadata(
    http_client: &reqwest::Client,
    metadata_url: &str,
    principal_id: &str,
) -> Option<String> {
    let url = format!(
        "{}/user-profile/{}?user_id={}",
        metadata_url.trim_end_matches('/'),
        principal_id,
        principal_id
    );

    match http_client.get(&url).send().await {
        Ok(resp) => {
            if !resp.status().is_success() {
                tracing::warn!(status = %resp.status(), "Metadata server returned error for user profile");
                return None;
            }
            match resp.json::<serde_json::Value>().await {
                Ok(json) => json
                    .get("Ok")
                    .and_then(|v| v.get("user_name"))
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string()),
                Err(e) => {
                    tracing::warn!(error = %e, "Failed to parse metadata user profile response");
                    None
                }
            }
        }
        Err(e) => {
            tracing::warn!(error = %e, "Failed to fetch username from metadata server");
            None
        }
    }
}

impl From<AIInfluencer> for InfluencerResponse {
    fn from(i: AIInfluencer) -> Self {
        Self {
            id: i.id,
            name: i.name,
            display_name: i.display_name,
            avatar_url: i.avatar_url,
            description: i.description,
            category: i.category,
            is_active: i.is_active,
            parent_principal_id: i.parent_principal_id,
            source: i.source,
            system_prompt: Some(moderation::strip_guardrails(&i.system_instructions)),
            created_at: i.created_at,
            conversation_count: i.conversation_count,
            message_count: i.message_count,
            starter_video_prompt: None,
        }
    }
}

type CachedJson<T> = ([(header::HeaderName, &'static str); 1], Json<T>);

/// List all influencers
#[utoipa::path(
    get,
    path = "/api/v1/influencers",
    params(PaginationParams),
    responses(
        (status = 200, body = ListInfluencersResponse, description = "Successful response"),
        (status = 422, body = ErrorBody, description = "Validation error")
    ),
    tag = "Influencers"
)]
pub async fn list_influencers(
    State(state): State<Arc<AppState>>,
    Query(params): Query<PaginationParams>,
) -> Result<CachedJson<ListInfluencersResponse>, AppError> {
    let repo = state.db.inf_repo();

    let limit = params.limit(50, 100);
    let offset = params.offset();

    let (influencers, total) = tokio::try_join!(repo.list_all(limit, offset), repo.count_all(),)?;

    Ok((
        [(header::CACHE_CONTROL, "public, max-age=300")],
        Json(ListInfluencersResponse {
            influencers: influencers
                .into_iter()
                .map(InfluencerResponse::from)
                .collect(),
            total,
            limit,
            offset,
        }),
    ))
}

/// List trending influencers
#[utoipa::path(
    get,
    path = "/api/v1/influencers/trending",
    params(PaginationParams),
    responses(
        (status = 200, body = ListTrendingInfluencersResponse, description = "Successful response"),
        (status = 422, body = ErrorBody, description = "Validation error")
    ),
    tag = "Influencers"
)]
pub async fn list_trending(
    State(state): State<Arc<AppState>>,
    Query(params): Query<PaginationParams>,
) -> Result<CachedJson<ListTrendingInfluencersResponse>, AppError> {
    let repo = state.db.inf_repo();

    let limit = params.limit(50, 100);
    let offset = params.offset();

    let (influencers, total) =
        tokio::try_join!(repo.list_trending(limit, offset), repo.count_trending(),)?;

    let influencers = influencers
        .into_iter()
        .map(|i| TrendingInfluencerResponse {
            id: i.id,
            name: i.name,
            display_name: i.display_name,
            avatar_url: i.avatar_url,
            description: i.description,
            category: i.category,
            is_active: i.is_active,
            created_at: i.created_at,
            conversation_count: i.conversation_count.unwrap_or(0),
            message_count: i.message_count.unwrap_or(0),
        })
        .collect();

    Ok((
        [(header::CACHE_CONTROL, "public, max-age=300")],
        Json(ListTrendingInfluencersResponse {
            influencers,
            total,
            limit,
            offset,
        }),
    ))
}

/// Get an influencer by ID
#[utoipa::path(
    get,
    path = "/api/v1/influencers/{influencer_id}",
    params(("influencer_id" = String, Path, description = "Influencer ID")),
    responses(
        (status = 200, body = InfluencerResponse),
        (status = 404, body = ErrorBody)
    ),
    tag = "Influencers"
)]
pub async fn get_influencer(
    State(state): State<Arc<AppState>>,
    Path(influencer_id): Path<String>,
) -> Result<CachedJson<InfluencerResponse>, AppError> {
    let repo = state.db.inf_repo();

    let influencer = repo
        .get_with_conversation_count(&influencer_id)
        .await?
        .ok_or_else(|| AppError::not_found(format!("Influencer '{influencer_id}' not found")))?;

    Ok((
        [(header::CACHE_CONTROL, "public, max-age=300")],
        Json(InfluencerResponse::from(influencer)),
    ))
}

/// Generate a system prompt from a user description
#[utoipa::path(
    post,
    path = "/api/v1/influencers/generate-prompt",
    request_body = GeneratePromptRequest,
    responses(
        (status = 200, body = SystemPromptResponse, description = "Successful response"),
        (status = 401, body = ErrorBody, description = "Unauthorized"),
        (status = 422, body = ErrorBody, description = "Validation error")
    ),
    tag = "Influencers",
    security(("BearerAuth" = []))
)]
pub async fn generate_prompt(
    State(state): State<Arc<AppState>>,
    _user: AuthenticatedUser,
    Json(body): Json<GeneratePromptRequest>,
) -> Result<Json<SystemPromptResponse>, AppError> {
    let instructions =
        CharacterGeneratorService::generate_system_instructions(&state.gemini, &body.prompt)
            .await?;

    Ok(Json(SystemPromptResponse {
        system_instructions: instructions,
    }))
}

/// Validate system instructions and generate influencer metadata
#[utoipa::path(
    post,
    path = "/api/v1/influencers/validate-and-generate-metadata",
    request_body = ValidateMetadataRequest,
    responses(
        (status = 200, body = GeneratedMetadataResponse, description = "Successful response"),
        (status = 401, body = ErrorBody, description = "Unauthorized"),
        (status = 422, body = ErrorBody, description = "Validation error")
    ),
    tag = "Influencers",
    security(("BearerAuth" = []))
)]
pub async fn validate_and_generate_metadata(
    State(state): State<Arc<AppState>>,
    _user: AuthenticatedUser,
    Json(body): Json<ValidateMetadataRequest>,
) -> Result<Json<GeneratedMetadataResponse>, AppError> {
    let result = CharacterGeneratorService::validate_and_generate_metadata(
        &state.gemini,
        &state.replicate,
        &body.system_instructions,
    )
    .await?;

    Ok(Json(result))
}

/// Create a new AI influencer
#[utoipa::path(
    post,
    path = "/api/v1/influencers/create",
    request_body = CreateInfluencerRequest,
    responses(
        (status = 200, body = InfluencerResponse, description = "Successful response"),
        (status = 401, body = ErrorBody, description = "Unauthorized"),
        (status = 409, body = ErrorBody, description = "Conflict"),
        (status = 422, body = ErrorBody, description = "Validation error")
    ),
    tag = "Influencers",
    security(("BearerAuth" = []))
)]
pub async fn create_influencer(
    State(state): State<Arc<AppState>>,
    user: AuthenticatedUser,
    Json(body): Json<CreateInfluencerRequest>,
) -> Result<Json<InfluencerResponse>, AppError> {
    // Validate request body
    body.validate()
        .map_err(|e| AppError::validation_error(format!("{e}")))?;

    let repo = state.db.inf_repo();

    // Check name uniqueness
    if let Some(_existing) = repo.get_by_name(&body.name).await? {
        return Err(AppError::conflict(format!(
            "Influencer name '{}' already exists",
            body.name
        )));
    }

    // Append moderation guardrails
    let system_instructions = moderation::with_guardrails(&body.system_instructions);

    // Generate greeting/suggestions if either is missing (matches Python behavior)
    let mut initial_greeting = body.initial_greeting.clone();
    let mut suggested_messages = body.suggested_messages.clone();

    if initial_greeting.is_none() || suggested_messages.is_empty() {
        match CharacterGeneratorService::generate_initial_greeting(
            &state.gemini,
            &body.display_name,
            &body.system_instructions,
        )
        .await
        {
            Ok((gen_greeting, gen_suggestions)) => {
                if initial_greeting.is_none() {
                    initial_greeting = Some(gen_greeting);
                }
                if suggested_messages.is_empty() {
                    suggested_messages = gen_suggestions;
                }
            }
            Err(e) => {
                tracing::error!(error = %e, "Failed to generate greeting");
                if initial_greeting.is_none() {
                    initial_greeting = Some(format!(
                        "Hey! I'm {}! How can I help you today?",
                        body.display_name
                    ));
                }
            }
        }
    }

    // Always use the authenticated user's ID (security: prevent override)
    let parent_principal_id = user.user_id.clone();

    let now = chrono::Utc::now().naive_utc();
    let influencer = AIInfluencer {
        id: body.bot_principal_id.clone(),
        name: body.name,
        display_name: body.display_name.clone(),
        avatar_url: body.avatar_url,
        description: body.description,
        category: body.category,
        system_instructions,
        personality_traits: body.personality_traits,
        initial_greeting,
        suggested_messages,
        is_active: InfluencerStatus::Active,
        is_nsfw: false, // enforced
        parent_principal_id: Some(parent_principal_id),
        source: Some("user-created-influencer".to_string()),
        created_at: now,
        updated_at: now,
        metadata: serde_json::json!({}),
        conversation_count: None,
        message_count: None,
    };

    repo.create(&influencer).await?;

    // Generate starter video prompt in parallel (best-effort)
    let starter_video_prompt = match CharacterGeneratorService::generate_starter_video_prompt(
        &state.gemini,
        &body.display_name,
        &body.system_instructions,
    )
    .await
    {
        Ok(prompt) => Some(prompt),
        Err(e) => {
            tracing::error!(error = %e, "Failed to generate starter video prompt");
            None
        }
    };

    let mut resp = InfluencerResponse::from(influencer);
    resp.starter_video_prompt = starter_video_prompt;

    Ok(Json(resp))
}

/// Update an influencer's system prompt
#[utoipa::path(
    patch,
    path = "/api/v1/influencers/{influencer_id}/system-prompt",
    params(("influencer_id" = String, Path, description = "Influencer ID")),
    request_body = UpdateSystemPromptRequest,
    responses(
        (status = 200, body = InfluencerResponse, description = "Successful response"),
        (status = 401, body = ErrorBody, description = "Unauthorized"),
        (status = 403, body = ErrorBody, description = "Forbidden"),
        (status = 404, body = ErrorBody, description = "Not found"),
        (status = 422, body = ErrorBody, description = "Validation error")
    ),
    tag = "Influencers",
    security(("BearerAuth" = []))
)]
pub async fn update_system_prompt(
    State(state): State<Arc<AppState>>,
    user: AuthenticatedUser,
    Path(influencer_id): Path<String>,
    Json(body): Json<UpdateSystemPromptRequest>,
) -> Result<Json<InfluencerResponse>, AppError> {
    let repo = state.db.inf_repo();

    let influencer = repo
        .get_by_id(&influencer_id)
        .await?
        .ok_or_else(|| AppError::not_found("Influencer not found"))?;

    // Only the owner can update
    if influencer.parent_principal_id.as_deref() != Some(&user.user_id) {
        return Err(AppError::forbidden(
            "Only the bot owner can update system prompt",
        ));
    }

    let instructions = moderation::with_guardrails(&body.system_instructions);
    repo.update_system_prompt(&influencer_id, &instructions)
        .await?;

    let updated = repo
        .get_by_id(&influencer_id)
        .await?
        .ok_or_else(|| AppError::not_found("Influencer not found"))?;

    Ok(Json(InfluencerResponse::from(updated)))
}

/// Generate a video prompt for subsequent bot videos
/// This endpoint creates an LTX-optimized video prompt with full context from the bot's system instructions
#[utoipa::path(
    post,
    path = "/api/v1/influencers/{influencer_id}/generate-video-prompt",
    params(("influencer_id" = String, Path, description = "Influencer ID")),
    request_body = GenerateVideoPromptRequest,
    responses(
        (status = 200, body = VideoPromptResponse, description = "Successful response"),
        (status = 401, body = ErrorBody, description = "Unauthorized"),
        (status = 404, body = ErrorBody, description = "Not found"),
        (status = 422, body = ErrorBody, description = "Validation error")
    ),
    tag = "Influencers",
    security(("BearerAuth" = []))
)]
pub async fn generate_video_prompt(
    State(state): State<Arc<AppState>>,
    _user: AuthenticatedUser,
    Path(influencer_id): Path<String>,
    Json(body): Json<GenerateVideoPromptRequest>,
) -> Result<Json<VideoPromptResponse>, AppError> {
    // Validate request body
    body.validate()
        .map_err(|e| AppError::validation_error(format!("{e}")))?;

    let repo = state.db.inf_repo();

    // Try to get the influencer - if not found, try to fetch profile from canister for main accounts
    let video_prompt = match repo.get_by_id(&influencer_id).await? {
        Some(influencer) => {
            // Bot context available - generate prompt with bot's system instructions
            let system_instructions = moderation::strip_guardrails(&influencer.system_instructions);
            CharacterGeneratorService::generate_subsequent_video_prompt(
                &state.gemini,
                &influencer.display_name,
                &system_instructions,
                &body.scene_description,
                influencer.avatar_url.as_deref(),
                body.reference_image_url.as_deref(),
            )
            .await?
        }
        None => {
            // Not a bot - fetch profile info for main user account
            let (avatar_url, username) = tokio::join!(
                fetch_user_profile_pic(&state.ic_agent, &influencer_id),
                fetch_username_from_metadata(
                    &state.http_client,
                    &state.settings.metadata_url,
                    &influencer_id
                )
            );

            // Use username from metadata, or fallback to short principal
            let display_name = username.unwrap_or_else(|| {
                format!("User {}", &influencer_id[..8.min(influencer_id.len())])
            });

            // Generate base prompt with avatar (if found) since main accounts don't have system instructions
            CharacterGeneratorService::generate_base_video_prompt(
                &state.gemini,
                &display_name,
                &body.scene_description,
                avatar_url.as_deref(),
                body.reference_image_url.as_deref(),
            )
            .await?
        }
    };

    Ok(Json(VideoPromptResponse {
        video_prompt,
        scene_description: body.scene_description,
    }))
}

/// Delete an influencer (soft delete) — owner only
#[utoipa::path(
    delete,
    path = "/api/v1/influencers/{influencer_id}",
    params(("influencer_id" = String, Path, description = "Influencer ID")),
    responses(
        (status = 200, body = InfluencerResponse, description = "Successful response"),
        (status = 401, body = ErrorBody, description = "Unauthorized"),
        (status = 403, body = ErrorBody, description = "Forbidden"),
        (status = 404, body = ErrorBody, description = "Not found")
    ),
    tag = "Influencers",
    security(("BearerAuth" = []))
)]
pub async fn delete_influencer(
    State(state): State<Arc<AppState>>,
    user: AuthenticatedUser,
    Path(influencer_id): Path<String>,
) -> Result<Json<InfluencerResponse>, AppError> {
    let repo = state.db.inf_repo();

    let influencer = repo
        .get_by_id(&influencer_id)
        .await?
        .ok_or_else(|| AppError::not_found("Influencer not found"))?;

    // Only the owner can delete
    if influencer.parent_principal_id.as_deref() != Some(&user.user_id) {
        return Err(AppError::forbidden(
            "Only the bot owner can delete this bot",
        ));
    }

    repo.soft_delete(&influencer_id).await?;

    let updated = repo
        .get_by_id(&influencer_id)
        .await?
        .ok_or_else(|| AppError::not_found("Influencer not found"))?;

    Ok(Json(InfluencerResponse::from(updated)))
}

/// Ban an influencer (admin only) — requires X-Admin-Key header
#[utoipa::path(
    delete,
    path = "/api/v1/admin/influencers/{influencer_id}",
    params(("influencer_id" = String, Path, description = "Influencer ID")),
    responses(
        (status = 200, body = InfluencerResponse, description = "Influencer banned"),
        (status = 401, body = ErrorBody, description = "Missing or invalid admin key"),
        (status = 404, body = ErrorBody, description = "Not found")
    ),
    tag = "Admin"
)]
pub async fn admin_ban_influencer(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Path(influencer_id): Path<String>,
) -> Result<Json<InfluencerResponse>, AppError> {
    let provided_key = headers
        .get("X-Admin-Key")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");

    let valid = state
        .settings
        .admin_key_to_delete_influencer
        .as_deref()
        .is_some_and(|key| provided_key == key);

    if !valid {
        return Err(AppError::unauthorized("Invalid or missing admin key"));
    }

    let repo = state.db.inf_repo();

    let influencer = repo
        .get_by_id_or_name(&influencer_id)
        .await?
        .ok_or_else(|| AppError::not_found("Influencer not found"))?;

    if let Err(e) = repo.ban(&influencer.id).await {
        state
            .google_chat
            .notify_influencer_ban_failed(&influencer.id, &e.to_string())
            .await;
        return Err(e.into());
    }

    state
        .google_chat
        .notify_influencer_banned(&influencer.id, &influencer.name)
        .await;

    Ok(Json(InfluencerResponse::from(influencer)))
}

/// Unban an influencer (admin only) — requires X-Admin-Key header
#[utoipa::path(
    post,
    path = "/api/v1/admin/influencers/{influencer_id}/unban",
    params(("influencer_id" = String, Path, description = "Influencer ID")),
    responses(
        (status = 200, body = InfluencerResponse, description = "Influencer unbanned"),
        (status = 401, body = ErrorBody, description = "Missing or invalid admin key"),
        (status = 404, body = ErrorBody, description = "Not found")
    ),
    tag = "Admin"
)]
pub async fn admin_unban_influencer(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Path(influencer_id): Path<String>,
) -> Result<Json<InfluencerResponse>, AppError> {
    let provided_key = headers
        .get("X-Admin-Key")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");

    let valid = state
        .settings
        .admin_key_to_delete_influencer
        .as_deref()
        .is_some_and(|key| provided_key == key);

    if !valid {
        return Err(AppError::unauthorized("Invalid or missing admin key"));
    }

    let repo = state.db.inf_repo();

    let influencer = repo
        .get_by_id_or_name(&influencer_id)
        .await?
        .ok_or_else(|| AppError::not_found("Influencer not found"))?;

    if let Err(e) = repo.unban(&influencer.id).await {
        state
            .google_chat
            .notify_influencer_unban_failed(&influencer.id, &e.to_string())
            .await;
        return Err(e.into());
    }

    state
        .google_chat
        .notify_influencer_unbanned(&influencer.id, &influencer.name)
        .await;

    Ok(Json(InfluencerResponse::from(influencer)))
}
